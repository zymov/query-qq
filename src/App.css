.App {
  padding: 50px 0;
  margin: 0 auto;
  width: 240px;
}

.title {
  font-size: 20px;
  margin-bottom: 10px;
  font-weight: bold;
}

.input {
  width: 100%;
  line-height: 24px;
  padding: 0 5px;
  margin-bottom: 20px;
  box-sizing: border-box;
}
.result {
  display: flex;
  align-items: center;
  padding: 10px;
  border: 1px solid #ddd;
  overflow: hidden;
  height: 60px;
}
.avatar {
  width: 60px;
  height: 60px;
}
.info {
  flex: 1 1;
  display: flex;
  padding: 10px;
  flex-direction: column;
  justify-content: center;
}
.name {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  flex-grow: 0;
  width: 138px;
}