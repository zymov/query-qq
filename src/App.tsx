import React, { useEffect, useState } from "react";
import "./App.css";
import { useDebounce } from "./hook";

type QQInfo = {
	name: string;
	qlogo: string;
	qq: string;
};

export function isValidQQ(qq: string): boolean {
	return /^[1-9]\d{4,11}$/.test(qq);
}

function App() {
	const [loading, setLoading] = useState(false);
	const [error, setError] = useState<string>("");
	const [data, setData] = useState<QQInfo | null>(null);
	const [search, setSearch] = useState("");

	const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		setSearch(e.target.value.trim());
	};

	const handleSearch = (value: string) => {
		if (value.length === 0) {
			setLoading(false);
			setData(null);
			setError("");
			return;
		}

		if (!isValidQQ(value)) {
			setData(null);
			setError("请输入正确的QQ号");
			return;
		}

		setError("");
		setLoading(true);
		fetch(`https://api.uomg.com/api/qq.info?qq=${value}`)
			.then((res) => res.json())
			.then((data) => {
				if (data.code !== 1) {
					setError(data.msg);
					setData(null);
				} else {
					setError("");
					setData({
						name: data.name,
						qlogo: data.qlogo,
						qq: data.qq,
					});
				}
			})
			.finally(() => setLoading(false));
	};

	const debouncedSearch = useDebounce(search, 500);

	useEffect(() => {
		handleSearch(debouncedSearch);
	}, [debouncedSearch]);

	return (
		<div className="App">
			<input
				className="input"
				type="text"
				placeholder="Search QQ"
				value={search}
				onChange={onChange}
			/>
			<div className="result">
				{loading && <div className="searching">Searching...</div>}
				{!loading && data && (
					<>
						<img
							className="avatar"
							src={data?.qlogo}
							alt="avatar"
						/>
						<div className="info">
							<div title={data?.name} className="name">
								{data?.name}
							</div>
							<div className="qq">{data?.qq}</div>
						</div>
					</>
				)}
				{error && <div className="error">{error}</div>}
			</div>
		</div>
	);
}

export default App;
