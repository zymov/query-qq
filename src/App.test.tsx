import App, { isValidQQ } from "./App";
import renderer from "react-test-renderer";

test("valid input", () => {
	expect(isValidQQ("123456789")).toBe(true);
	expect(isValidQQ("a123456789")).toBe(false);
	expect(isValidQQ("01234")).toBe(false);
	expect(isValidQQ("abc")).toBe(false);
	expect(isValidQQ("  ")).toBe(false);
});

test("snapshot", () => {
	const tree: any = renderer.create(<App />).toJSON();
	expect(tree).toMatchSnapshot();
});

test("fetch with wrong qq value", async () => {
	const data = await fetch(`https://api.uomg.com/api/qq.info?qq=123`).then(
		(res) => res.json()
	);
	expect(data.code).not.toBe(1);
});

test("fetch with right qq value", async () => {
	const data = await fetch(`https://api.uomg.com/api/qq.info?qq=123456`).then(
		(res) => res.json()
	);
	expect(data.code).toBe(1);
	expect(data.qq).toBe("123456");
});
