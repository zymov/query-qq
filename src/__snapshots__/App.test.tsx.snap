// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
<div
  className="App"
>
  <input
    className="input"
    onChange={[Function]}
    placeholder="Search QQ"
    type="text"
    value=""
  />
  <div
    className="result"
  />
</div>
`;

exports[`rendr 1`] = `
<div
  className="App"
>
  <input
    className="input"
    onChange={[Function]}
    placeholder="Search QQ"
    type="text"
    value=""
  />
  <div
    className="result"
  />
</div>
`;

exports[`snapshot 1`] = `
<div
  className="App"
>
  <input
    className="input"
    onChange={[Function]}
    placeholder="Search QQ"
    type="text"
    value=""
  />
  <div
    className="result"
  />
</div>
`;
